### ISSUE TITLE
Edit Commit Message

### ISSUE THREAD

**Issue by JosephTLyons** [desktop/desktop#5219]
**Is your feature request related to a problem? Please describe.**
No, it is just a convenient feature I'd love to have

**Describe the solution you'd like**
Right click on a commit and see the option for "edit git commit message" show up.

**Describe alternatives you've considered**
Using the command line or another Git Client are the only other options:

<img width="470" alt="screen shot 2018-07-23 at 3 08 59 am" src="https://user-images.githubusercontent.com/19867440/43062258-41876df8-8e26-11e8-9d8b-034631e16499.png">

**Teachability, Documentation, Adoption, Migration Strategy**
This is a great feature for when we hastily make a commit message that isn't as accurate as we would like or make a message that has a typo.

**Comment by j-f1** [comment]
This is what the “Undo” button is for — you can click it, edit the commit message, then re-commit.

**Comment by JosephTLyons** [comment]
Correct, but it isn’t possible to rename a commit other than the most recent, without multiple undos, correct?

**Comment by shiftkey** [comment]
> Correct, but it isn’t possible to rename a commit other than the most recent, without multiple undos, correct?

This is correct, and when you undo a commit the changes (and commit message) are added to the working directory - so this means your commits are merged into one by the time you get to the one you want to change.

**Comment by aminya** [comment]
## Rewording last commit's comment
There should be a command in the software that we can edit the comment of our last commit. It should do the following git commands:
```
git commit --amend -m "title" -m "

Your comment here
use enter for different lines

"
```
Then
```
git push --force
```
_P.S: I am not familiar with TypeScript or React, otherwise, I would have helped to implement some new features._

**Comment by j-f1** [comment]
@aminya You can already do this by undoing the most recent commit, rewording the message, then committing again.

**Comment by protoEvangelion** [comment]
Can you rename a commit message deep in history though?

**Comment by aminya** [comment]
> 
> 
> @aminya You can already do this by undoing the most recent commit, rewording the message, then committing again.

@j-f1 Now it is possible. At that time there was no undo, we only had revert


> 
> 
> Can you rename a commit message deep in history though?

No, it is not possible

**Comment by j-f1** [comment]
> At that time there was no undo, we only had revert

I’m pretty sure undo has existed for most of Desktop’s history.

**Comment by JosephTLyons** [comment]
> > At that time there was no undo, we only had revert
> 
> I’m pretty sure undo has existed for most of Desktop’s history.

Right.  Undo was available when I made this issue.

**Comment by aminya** [comment]
> 
> 
> > > At that time there was no undo, we only had revert
> > 
> > 
> > I’m pretty sure undo has existed for most of Desktop’s history.
> 
> Right. Undo was available when I made this issue.

No, my problem is that when you push the commit the undo get disapread. I need to use git bash to undo and then force push!

**Comment by JosephTLyons** [comment]
> > > > At that time there was no undo, we only had revert
> > > 
> > > 
> > > I’m pretty sure undo has existed for most of Desktop’s history.
> > 
> > 
> > Right. Undo was available when I made this issue.
> 
> No, my problem is that when you push the commit the undo get disapread. I need to use git bash to undo and then force push!

This is what I do as well.  There are two reasons in my mind why GitHub Desktop doesn't have this feature:

1) They haven't gotten around to it.
2) Using git `reset --soft HEAD~1` and `git push -f` can cause issues for other people who have your latest commits pulled down.

**Comment by DCzajkowski** [comment]
In my opinion this should be available only on unpushed commits, because:
- it prevents people unfamiliar with Git from getting unexpected results
- it teaches good Git practices (no force pushing)
- it eliminates a situation, where an action in GH Desktop forces one to use a terminal anyway (to force push)

For now, if you want to edit a commit message back in the history you need to `git rebase -i HEAD~n`, where n is the _number of commits you want to go back **+ 1**_, on the commit you want to edit, pick the `reword` action, save and exit the editor, change its message, save and exit the editor.

**Comment by aminya** [comment]
>     * it teaches good Git practices (no force pushing)

I don't see limiting people as a form of teaching. This approach punishes the developer and the repo maintainers and produces a mess when someone makes a mistake. 

**While Git already has solutions for human's mistakes, GitHub Desktop doesn't.**

**Comment by stlehmann** [comment]
I would also appreciate that functionality. SourceTree also provides this feature by using interactive rebase. I often find it very helpful to reorganize or rename the commits before pushing them to origin.

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/16130.

**Comment by 0xdevalias** [comment]
> In my opinion this should be available only on unpushed commits, because:
> 
> * it prevents people unfamiliar with Git from getting unexpected results
> * it teaches good Git practices (no force pushing)
> * it eliminates a situation, where an action in GH Desktop forces one to use a terminal anyway (to force push)

I would suggest limiting this to unpushed commits isn't the right choice here.

I didn't see this mentioned above anywhere, but (at least as of today in 2023), on the latest commit, you can use the 'amend commit' option:

![image](https://github.com/desktop/desktop/assets/753891/561d324b-68be-4f07-95f7-8cfba9a562fd)

This works even if it has already been pushed, but warns the user of the repercussions first:

![image](https://github.com/desktop/desktop/assets/753891/6904a289-c96b-421c-906f-eef325843388)

That's a good start, but it only works on the most recent commit. If we try to 'ammend' on anything deeper, there is no option:

![image](https://github.com/desktop/desktop/assets/753891/ee99c381-f5e2-4536-a759-dd54b84bb63b)

Outside of GitHub Desktop, this would be achieved using interactive rebase (Ref: [1](https://github.com/desktop/desktop/issues/5219#issuecomment-614799420), [2](https://github.com/desktop/desktop/issues/5219#issuecomment-620475330))

Given there is already an existing warning dialog about a force push being required, it doesn't seem like there should be much harm from allowing it on an earlier commit.

One potential concern off the top of my head would be how to handle conflicts during the interactive rebase (assuming this doesn't work already). One potential workaround for that could be to restrict the 'ammend commit' functionality on older commits to only allow amending the commit message itself (changing the menu option to something like 'ammend commit message' or 'edit commit message'); but otherwise using most of the same underlying functionality (and warning message) of the existing 'ammend'.

---

**Edit:** The fact that there is already a 'reorder commit' option, even on commits beyond the first one makes me think there must already be decent support for interactive rebase in the internals:

![image](https://github.com/desktop/desktop/assets/753891/c2634463-dc82-4f84-9712-59beda6e7d7f)

![image](https://github.com/desktop/desktop/assets/753891/036e6588-23ee-47bc-ad39-e8b822fcb2f6)

Interestingly, even for a pushed branch, this 'reorder commit' doesn't actually show a warning about force push being required (at least, not before it has been moved; I didn't try and see if it popped up after I tried)

**Edit 2:** Created a throwaway branch to test this, and it seems the warning is shown when you hit enter to confirm the reorder:

![image](https://github.com/desktop/desktop/assets/753891/5fad832c-df71-4fff-a588-f57ff52dcef7)

**Edit 3:** See also this issue RE: interactive rebase:

- https://github.com/desktop/desktop/issues/12354

**Comment by 0xdevalias** [comment]
Seems this was also requested here:

> The reordering and squashing features are very awesome, thank you for adding them! I think at the moment there are mainly two features missing to get a very good rebase experience using GitHub Desktop:
> 
> 1. Being able to drop (i. e. delete) commits from the history
> 2. **Being able to change a commit message without squashing**
> 
> Are you planning on adding in these as well?
> 
> _Originally posted by @JKRhb in https://github.com/desktop/desktop/issues/12354#issuecomment-874366125_

**Comment by 0xdevalias** [comment]
RE: https://github.com/desktop/desktop/issues/5219#issuecomment-1858996516

> The fact that there is already a 'reorder commit' option, even on commits beyond the first one makes me think there must already be decent support for interactive rebase in the internals

Dove into the source code for a bit of a search, and here are some of the lib files related to `rebaseInteractive`. There may be more than this, and I didn't look too deeply as to where/how these are used by the UI itself, but it could be a good starting place for anyone wanting to look into this stuff deeper:

> From a quick search in the code, I found this git lib function for `squash` that sounds like it's probably used for the squash functionality (via interactive rebase):
> 
> https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/lib/git/squash.ts#L10-L40
> 
> It then calls into `rebaseInteractive`:
> 
> https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/lib/git/squash.ts#L148-L156
> 
> `rebaseInteractive` is defined here:
> 
> https://github.com/desktop/desktop/blob/development/app/src/lib/git/rebase.ts#L533-L556
> 
> That seems to make the actual call to `git rebase -i` here, and doesn't look like it currently uses `--autostash` at all:
> 
> https://github.com/desktop/desktop/blob/development/app/src/lib/git/rebase.ts#L582-L594
> 
> _Originally posted by @0xdevalias in https://github.com/desktop/desktop/issues/12759#issuecomment-1859055247_

---

RE: https://github.com/desktop/desktop/issues/5219#issuecomment-1858996516

> I didn't see this mentioned above anywhere, but (at least as of today in 2023), on the latest commit, you can use the 'amend commit' option

> That's a good start, but it only works on the most recent commit. If we try to 'ammend' on anything deeper, there is no option

You can see where the 'Amend Commit' menu item is defined within `CommitList` here, which then calls into `onAmendCommit` passed in from the props:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/history/commit-list.tsx#L84

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/history/commit-list.tsx#L654-L659

It's gated by `canBeAmended`, which seems to make this only apply to the first item by the `&& row === 0` in it's definition:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/history/commit-list.tsx#L632

I haven't deeply explore the repercussions of this idea yet, but, at least in theory, this could be made to apply to all commits by changing it to just be:

```diff
- const canBeAmended = this.props.canAmendCommits === true && row === 0
+ const canBeAmended = this.props.canAmendCommits === true
```

Obviously the impact of this would need to be explored further, and tested thoroughly; but if anyone wants to look into it deeper, that would be a good place to start.

You can also see a more involved deep dive into the code structure about how `CommitList` actually gets called/what gets passed through to it (all the way up to the `App` component) in this comment:

- https://github.com/desktop/desktop/issues/9791#issuecomment-1859049098
  - **tl;dr:** [`App`](https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/app.tsx#L3268-L3312) -> [`RepositoryView`](https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/repository.tsx#L259-L311) -> [`CompareSidebar`](https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/history/compare.tsx#L237-L252) -> [`CommitList`](https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/history/commit-list.tsx#L187-L191) -> [`getContextMenuForSingleCommit`](https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/history/commit-list.tsx#L624-L769)

**Comment by mohdwaseem** [comment]
It would be very helpful if we have the option to amend the older un-pushed commit message.

**Comment by ghost** [comment]
+1

**Comment by rozbeste** [comment]
I just worked out a way to do this in Github Desktop.(Only do this if  you know you are the only one working on this branch as it requires a force push)
1. Create a new commit with maybe an extra line break in a file that was changed in the commit where you want to change the message.
2. Commit to local branch
3. Go to History tab. Select your new commit, hold ctrl and also select the commit where you want to change the message. Right-click. Squash 2 commits. This will merge your new commit with the older commit and it allows you to change the message. You will then see that you need to force-push.
